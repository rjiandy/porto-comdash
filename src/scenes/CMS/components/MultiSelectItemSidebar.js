// @flow
import React, {Component} from 'react';
import autobind from 'class-autobind';
import {Checkbox} from 'material-ui';
import {StyleSheet, Touchable} from 'react-primitives';
import {
  View,
  Button,
  TextInput,
  Text,
  MaterialIcon,
  Icon,
  ScrollView,
} from '../../../general/components/coreUIComponents';

import {
  TEXT_COLOR,
  THEME_COLOR,
  LIGHT_GREY,
  ALTERNATIVE_GREY,
  MEDIUM_GREY,
  MAIN_BLUE,
} from '../../../general/constants/colors';
import {
  SMALL_FONT_SIZE,
  DEFAULT_FONT_FAMILY,
} from '../../../general/constants/text';

type State = {
  selectedItems: Set<string>;
  filterValue: string;
  isAssignedFilterActive: boolean;
};

type ItemID = string;

type Item = {
  id: ItemID;
  name: string;
};
type Props = {
  icon?: string;
  searchPlaceholder: string;
  noItemFound: string;
  itemList: Array<Item>;
  selectedItems: Array<ItemID>;
  onConfirm: (selectedItems: Array<ItemID>) => void;
  toggleSidebar: (command: 'open' | 'close', data?: Array<ItemID>) => void;
  onAddData?: () => void;
  addDataButtonLabel?: string;
};

export type MultiSelectItemSidebarProps = Props;

export default class MultiSelectItemSidebar extends Component {
  state: State;
  props: Props;

  constructor() {
    super(...arguments);
    autobind(this);

    this.state = {
      selectedItems: this._getSelectedItems(this.props.selectedItems),
      filterValue: '',
      isAssignedFilterActive: false,
    };
  }

  componentWillReceiveProps(newProps: Props) {
    let oldProps = this.props;
    if (newProps.selectedItems.length !== oldProps.selectedItems.length) {
      this.setState({
        selectedItems: this._getSelectedItems(newProps.selectedItems),
      });
    }
  }
  _isTickAll(itemList: Array<Item>) {
    let {selectedItems} = this.state;
    for (let {id} of itemList) {
      if (!selectedItems.has(id)) {
        return false;
      }
    }
    if (itemList.length === 0) {
      return false;
    }
    return true;
  }
  _tickAll(itemList: Array<Item>) {
    let {selectedItems} = this.state;
    let newSelectedItems = new Set(selectedItems);
    let isTickAll = this._isTickAll(itemList);
    let isAssignedFilterActive = false;
    if (!isTickAll) {
      for (let {id} of itemList) {
        if (!selectedItems.has(id)) {
          newSelectedItems.add(id);
        }
      }
      isAssignedFilterActive = true;
    } else {
      for (let {id} of itemList) {
        newSelectedItems.delete(id);
      }
    }
    this.setState({
      selectedItems: newSelectedItems,
      isAssignedFilterActive,
    });
  }
  render() {
    let {selectedItems, filterValue, isAssignedFilterActive} = this.state;
    let {
      searchPlaceholder,
      itemList,
      toggleSidebar,
      onAddData,
      addDataButtonLabel,
      noItemFound,
      icon,
    } = this.props;

    let filteredItemList = itemList.filter(
      (item) =>
        item.name.toLowerCase().includes(filterValue.toLowerCase()) &&
        (isAssignedFilterActive
          ? selectedItems.has(item.id)
          : !selectedItems.has(item.id)),
    );

    return (
      <View style={styles.container}>
        <View style={{paddingHorizontal: 11, marginBottom: 10}}>
          <TextInput
            fullWidth
            hintText={searchPlaceholder}
            value={filterValue}
            onTextChange={(newFilterValue) =>
              this.setState({filterValue: newFilterValue})}
            icon="search"
            iconStyle={{color: LIGHT_GREY}}
          />
        </View>
        <View style={styles.itemsHeader}>
          <View style={{width: 40}}>
            <Checkbox
              uncheckedIcon={
                <MaterialIcon
                  name="check-box-outline-blank"
                  color={LIGHT_GREY}
                />
              }
              checkedIcon={
                <MaterialIcon name="check-box" style={{color: THEME_COLOR}} />
              }
              checked={this._isTickAll(filteredItemList)}
              onCheck={() => this._tickAll(filteredItemList)}
              style={styles.checkboxIcon}
            />
          </View>
          <View style={[styles.flexRow, styles.switchContainer]}>
            <Touchable
              onPress={() => this.setState({isAssignedFilterActive: true})}
            >
              <View style={styles.switchItemContainer}>
                <Text
                  style={{
                    fontSize: 10,
                    fontWeight: 'bold',
                    color: isAssignedFilterActive ? MAIN_BLUE : MEDIUM_GREY,
                  }}
                >
                  ASSIGNED
                </Text>
              </View>
            </Touchable>
            <Touchable
              onPress={() => this.setState({isAssignedFilterActive: false})}
            >
              <View style={styles.switchItemContainer}>
                <Text
                  style={{
                    fontSize: 10,
                    fontWeight: 'bold',
                    color: isAssignedFilterActive ? MEDIUM_GREY : MAIN_BLUE,
                  }}
                >
                  UNASSIGNED
                </Text>
              </View>
            </Touchable>
          </View>
        </View>
        <View style={styles.groupList}>
          {filteredItemList.length === 0 ? (
            <View style={styles.switchItemContainer}>
              <Text style={{textAlign: 'center', color: LIGHT_GREY}}>
                {noItemFound}
              </Text>
            </View>
          ) : (
            <ScrollView>
              {filteredItemList.map((item, index) => {
                return (
                  <View style={styles.rowItem} key={index}>
                    <Checkbox
                      uncheckedIcon={
                        <MaterialIcon
                          name="check-box-outline-blank"
                          color={LIGHT_GREY}
                        />
                      }
                      checkedIcon={
                        <MaterialIcon
                          name="check-box"
                          style={{color: THEME_COLOR}}
                        />
                      }
                      checked={selectedItems.has(item.id)}
                      onCheck={() => this._selectItem(item.id)}
                      style={StyleSheet.flatten([
                        styles.checkboxIcon,
                        {width: 40},
                      ])}
                    />
                    <View style={styles.checklistContent}>
                      <View
                        style={{
                          width: 20,
                          height: 20,
                          marginHorizontal: 8,
                        }}
                      >
                        <Icon name={icon || 'user'} />
                      </View>
                      <Text>{item.name}</Text>
                    </View>
                  </View>
                );
              })}
            </ScrollView>
          )}
        </View>
        <View style={{marginTop: 10}}>
          {onAddData && addDataButtonLabel ? (
            <Button
              secondary
              label={addDataButtonLabel}
              style={styles.button}
              onPress={onAddData}
            />
          ) : null}

          <View style={styles.bottomButtons}>
            <Button
              secondary
              label="Cancel"
              style={styles.button}
              onPress={() =>
                toggleSidebar('close', [...this.state.selectedItems.values()])}
            />
            <Button
              primary
              label="Confirm"
              style={StyleSheet.flatten([styles.button, {marginLeft: 15}])}
              onPress={() => this._onConfirm()}
            />
          </View>
        </View>
      </View>
    );
  }

  _getSelectedItems(selectedItems: Array<ItemID>) {
    return new Set(
      this.props.itemList
        .filter((item) => selectedItems.includes(item.id))
        .map((item) => item.id),
    );
  }

  _selectItem(item: string) {
    let {selectedItems} = this.state;
    if (selectedItems.has(item)) {
      selectedItems.delete(item);
    } else {
      selectedItems.add(item);
    }
    this.setState({selectedItems});
  }

  _onConfirm() {
    let {onConfirm, toggleSidebar} = this.props;
    let {selectedItems} = this.state;
    onConfirm(Array.from(selectedItems.values()));
    toggleSidebar('close', []);
  }
}

const styles = {
  container: {
    flex: 1,
  },
  flexRow: {
    flexDirection: 'row',
  },
  switchItemContainer: {
    flex: 1,
    justifyContent: 'center',
    alignItems: 'center',
  },
  switchContainer: {
    height: '100%',
    width: '80%',
    paddingVertical: 8,
  },
  itemsHeader: {
    height: 46,
    width: '90%',
    alignItems: 'center',
    marginHorizontal: 10,
    marginBottom: 6,
    borderBottomWidth: 1,
    borderBottomColor: ALTERNATIVE_GREY,
    flexDirection: 'row',
  },
  rowItem: {
    width: '90%',
    marginHorizontal: 10,
    marginBottom: 16,
    flexDirection: 'row',
    alignItems: 'center',
  },
  groupList: {
    flex: 1,
    paddingTop: 10,
  },
  bottomButtons: {
    flexDirection: 'row',
    marginTop: 10,
  },
  button: {
    flex: 1,
  },
  checkbox: {
    paddingLeft: 12,
    marginBottom: 10,
    marginHorizontal: 10,
    alignSelf: 'stretch',
    height: 42,
  },
  checkboxIcon: {
    color: THEME_COLOR,
    width: 20,
  },
  checklistContent: {
    flexDirection: 'row',
    borderWidth: 1,
    borderRadius: 2,
    borderColor: LIGHT_GREY,
    width: '80%',
    overflow: 'scroll', // TODO: Improve this styling later
    alignItems: 'center',
    paddingVertical: 10,
    paddingRight: 4,
  },
  checkboxLabel: {
    fontFamily: DEFAULT_FONT_FAMILY,
    fontSize: SMALL_FONT_SIZE,
    color: TEXT_COLOR,
  },
};
