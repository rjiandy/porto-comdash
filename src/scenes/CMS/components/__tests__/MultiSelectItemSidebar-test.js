// @flow

import React from 'react';
import {shallow} from 'enzyme';
import toJSON from 'enzyme-to-json';
// import renderer from 'react-test-renderer';

import MultiSelectItemSidebar from '../MultiSelectItemSidebar';

jest.mock('../../../../general/components/coreUIComponents', () => ({
  Button: 'Button',
  View: 'View',
  TextInput: 'TextInput',
  ScrollView: 'ScrollView',
}));
jest.mock('material-ui', () => ({
  Checkbox: 'Checkbox',
}));
jest.mock('react-primitives', () => ({
  Touchable: 'Touchable',
  StyleSheet: {
    create: () => {},
    flatten: () => {},
  },
}));

it('should render a Multi Selection Item Sidebar with correct props', () => {
  let itemList = [
    {id: 'item1', name: 'item 1'},
    {id: 'item2', name: 'item 2'},
    {id: 'item3', name: 'item 3'},
  ];
  let selectedItems = ['item1'];
  let toggleSidebar = jest.fn();
  let onConfirm = jest.fn((newSelectedItems) => newSelectedItems);
  let component = shallow(
    <MultiSelectItemSidebar
      id="id"
      noItemFound="no item found"
      searchPlaceholder="search"
      itemList={itemList}
      selectedItems={selectedItems}
      onConfirm={onConfirm}
      toggleSidebar={toggleSidebar}
    />
  );
  expect(toJSON(component)).toMatchSnapshot();
});

it('should render a Multi Selection Item Sidebar with add data button', () => {
  let itemList = [
    {id: 'item1', name: 'item 1'},
    {id: 'item2', name: 'item 2'},
    {id: 'item3', name: 'item 3'},
  ];
  let selectedItems = ['item1'];
  let toggleSidebar = jest.fn();
  let onAddData = jest.fn();
  let onConfirm = jest.fn((newSelectedItems) => newSelectedItems);
  let component = shallow(
    <MultiSelectItemSidebar
      id="id"
      noItemFound="no item found"
      searchPlaceholder="search"
      itemList={itemList}
      selectedItems={selectedItems}
      onConfirm={onConfirm}
      toggleSidebar={toggleSidebar}
      addDataButtonLabel="Add More Data"
      onAddData={onAddData}
    />
  );
  expect(toJSON(component)).toMatchSnapshot();
});
