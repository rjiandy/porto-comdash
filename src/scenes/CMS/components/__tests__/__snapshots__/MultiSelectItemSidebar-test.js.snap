// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render a Multi Selection Item Sidebar with add data button 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 10,
        "paddingHorizontal": 11,
      }
    }
  >
    <TextInput
      fullWidth={true}
      hintText="search"
      icon="search"
      iconStyle={
        Object {
          "color": "#E4EBE6",
        }
      }
      onTextChange={[Function]}
      value=""
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "#EBEBEB",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "height": 46,
        "marginBottom": 6,
        "marginHorizontal": 10,
        "width": "90%",
      }
    }
  >
    <View
      style={
        Object {
          "width": 40,
        }
      }
    >
      <Checkbox
        checked={false}
        checkedIcon={
          <undefined
            name="check-box"
            style={
              Object {
                "color": "#4A90E2",
              }
            }
          />
        }
        onCheck={[Function]}
        style={
          Object {
            "color": "#4A90E2",
            "width": 20,
          }
        }
        uncheckedIcon={
          <undefined
            color="#E4EBE6"
            name="check-box-outline-blank"
          />
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          Object {
            "height": "100%",
            "paddingVertical": 8,
            "width": "80%",
          },
        ]
      }
    >
      <Touchable
        onPress={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <undefined
            style={
              Object {
                "color": "#CCC",
                "fontSize": 10,
                "fontWeight": "bold",
              }
            }
          >
            ASSIGNED
          </undefined>
        </View>
      </Touchable>
      <Touchable
        onPress={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <undefined
            style={
              Object {
                "color": "#4A90E2",
                "fontSize": 10,
                "fontWeight": "bold",
              }
            }
          >
            UNASSIGNED
          </undefined>
        </View>
      </Touchable>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "paddingTop": 10,
      }
    }
  >
    <ScrollView>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 16,
            "marginHorizontal": 10,
            "width": "90%",
          }
        }
      >
        <Checkbox
          checked={false}
          checkedIcon={
            <undefined
              name="check-box"
              style={
                Object {
                  "color": "#4A90E2",
                }
              }
            />
          }
          onCheck={[Function]}
          uncheckedIcon={
            <undefined
              color="#E4EBE6"
              name="check-box-outline-blank"
            />
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#E4EBE6",
              "borderRadius": 2,
              "borderWidth": 1,
              "flexDirection": "row",
              "overflow": "scroll",
              "paddingRight": 4,
              "paddingVertical": 10,
              "width": "80%",
            }
          }
        >
          <View
            style={
              Object {
                "height": 20,
                "marginHorizontal": 8,
                "width": 20,
              }
            }
          >
            <undefined
              name="user"
            />
          </View>
          <undefined>
            item 2
          </undefined>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 16,
            "marginHorizontal": 10,
            "width": "90%",
          }
        }
      >
        <Checkbox
          checked={false}
          checkedIcon={
            <undefined
              name="check-box"
              style={
                Object {
                  "color": "#4A90E2",
                }
              }
            />
          }
          onCheck={[Function]}
          uncheckedIcon={
            <undefined
              color="#E4EBE6"
              name="check-box-outline-blank"
            />
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#E4EBE6",
              "borderRadius": 2,
              "borderWidth": 1,
              "flexDirection": "row",
              "overflow": "scroll",
              "paddingRight": 4,
              "paddingVertical": 10,
              "width": "80%",
            }
          }
        >
          <View
            style={
              Object {
                "height": 20,
                "marginHorizontal": 8,
                "width": 20,
              }
            }
          >
            <undefined
              name="user"
            />
          </View>
          <undefined>
            item 3
          </undefined>
        </View>
      </View>
    </ScrollView>
  </View>
  <View
    style={
      Object {
        "marginTop": 10,
      }
    }
  >
    <Button
      label="Add More Data"
      onPress={[Function]}
      secondary={true}
      style={
        Object {
          "flex": 1,
        }
      }
    />
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 10,
        }
      }
    >
      <Button
        label="Cancel"
        onPress={[Function]}
        secondary={true}
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <Button
        label="Confirm"
        onPress={[Function]}
        primary={true}
      />
    </View>
  </View>
</View>
`;

exports[`should render a Multi Selection Item Sidebar with correct props 1`] = `
<View
  style={
    Object {
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "marginBottom": 10,
        "paddingHorizontal": 11,
      }
    }
  >
    <TextInput
      fullWidth={true}
      hintText="search"
      icon="search"
      iconStyle={
        Object {
          "color": "#E4EBE6",
        }
      }
      onTextChange={[Function]}
      value=""
    />
  </View>
  <View
    style={
      Object {
        "alignItems": "center",
        "borderBottomColor": "#EBEBEB",
        "borderBottomWidth": 1,
        "flexDirection": "row",
        "height": 46,
        "marginBottom": 6,
        "marginHorizontal": 10,
        "width": "90%",
      }
    }
  >
    <View
      style={
        Object {
          "width": 40,
        }
      }
    >
      <Checkbox
        checked={false}
        checkedIcon={
          <undefined
            name="check-box"
            style={
              Object {
                "color": "#4A90E2",
              }
            }
          />
        }
        onCheck={[Function]}
        style={
          Object {
            "color": "#4A90E2",
            "width": 20,
          }
        }
        uncheckedIcon={
          <undefined
            color="#E4EBE6"
            name="check-box-outline-blank"
          />
        }
      />
    </View>
    <View
      style={
        Array [
          Object {
            "flexDirection": "row",
          },
          Object {
            "height": "100%",
            "paddingVertical": 8,
            "width": "80%",
          },
        ]
      }
    >
      <Touchable
        onPress={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <undefined
            style={
              Object {
                "color": "#CCC",
                "fontSize": 10,
                "fontWeight": "bold",
              }
            }
          >
            ASSIGNED
          </undefined>
        </View>
      </Touchable>
      <Touchable
        onPress={[Function]}
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "flex": 1,
              "justifyContent": "center",
            }
          }
        >
          <undefined
            style={
              Object {
                "color": "#4A90E2",
                "fontSize": 10,
                "fontWeight": "bold",
              }
            }
          >
            UNASSIGNED
          </undefined>
        </View>
      </Touchable>
    </View>
  </View>
  <View
    style={
      Object {
        "flex": 1,
        "paddingTop": 10,
      }
    }
  >
    <ScrollView>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 16,
            "marginHorizontal": 10,
            "width": "90%",
          }
        }
      >
        <Checkbox
          checked={false}
          checkedIcon={
            <undefined
              name="check-box"
              style={
                Object {
                  "color": "#4A90E2",
                }
              }
            />
          }
          onCheck={[Function]}
          uncheckedIcon={
            <undefined
              color="#E4EBE6"
              name="check-box-outline-blank"
            />
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#E4EBE6",
              "borderRadius": 2,
              "borderWidth": 1,
              "flexDirection": "row",
              "overflow": "scroll",
              "paddingRight": 4,
              "paddingVertical": 10,
              "width": "80%",
            }
          }
        >
          <View
            style={
              Object {
                "height": 20,
                "marginHorizontal": 8,
                "width": 20,
              }
            }
          >
            <undefined
              name="user"
            />
          </View>
          <undefined>
            item 2
          </undefined>
        </View>
      </View>
      <View
        style={
          Object {
            "alignItems": "center",
            "flexDirection": "row",
            "marginBottom": 16,
            "marginHorizontal": 10,
            "width": "90%",
          }
        }
      >
        <Checkbox
          checked={false}
          checkedIcon={
            <undefined
              name="check-box"
              style={
                Object {
                  "color": "#4A90E2",
                }
              }
            />
          }
          onCheck={[Function]}
          uncheckedIcon={
            <undefined
              color="#E4EBE6"
              name="check-box-outline-blank"
            />
          }
        />
        <View
          style={
            Object {
              "alignItems": "center",
              "borderColor": "#E4EBE6",
              "borderRadius": 2,
              "borderWidth": 1,
              "flexDirection": "row",
              "overflow": "scroll",
              "paddingRight": 4,
              "paddingVertical": 10,
              "width": "80%",
            }
          }
        >
          <View
            style={
              Object {
                "height": 20,
                "marginHorizontal": 8,
                "width": 20,
              }
            }
          >
            <undefined
              name="user"
            />
          </View>
          <undefined>
            item 3
          </undefined>
        </View>
      </View>
    </ScrollView>
  </View>
  <View
    style={
      Object {
        "marginTop": 10,
      }
    }
  >
    <View
      style={
        Object {
          "flexDirection": "row",
          "marginTop": 10,
        }
      }
    >
      <Button
        label="Cancel"
        onPress={[Function]}
        secondary={true}
        style={
          Object {
            "flex": 1,
          }
        }
      />
      <Button
        label="Confirm"
        onPress={[Function]}
        primary={true}
      />
    </View>
  </View>
</View>
`;
